selfdrive/ui/sound.cc

```diff
@@ -1,91 +1,125 @@
-#include <stdlib.h>
-#include "sound.hpp"
 
+#include "sound.hpp"
+#include <assert.h>
+#include <stdlib.h>
 #include "common/swaglog.h"
+#include "common/timing.h"
+
+#define ReturnOnError(func, msg) \
+  if ((func) != SL_RESULT_SUCCESS) { LOGW(msg); return false; }
+
+std::map<AudibleAlert, const char*> sound_map{
+  {AudibleAlert::CHIME_DISENGAGE, "../assets/sounds/disengaged.wav"},
+  {AudibleAlert::CHIME_ENGAGE, "../assets/sounds/engaged.wav"},
+  {AudibleAlert::CHIME_WARNING1, "../assets/sounds/warning_1.wav"},
+  {AudibleAlert::CHIME_WARNING2, "../assets/sounds/warning_2.wav"},
+  {AudibleAlert::CHIME_WARNING2_REPEAT, "../assets/sounds/warning_2.wav"},
+  {AudibleAlert::CHIME_WARNING_REPEAT, "../assets/sounds/warning_repeat.wav"},
+  {AudibleAlert::CHIME_ERROR, "../assets/sounds/error.wav"},
+  {AudibleAlert::CHIME_PROMPT, "../assets/sounds/error.wav"},
+};
 
-typedef struct {
-  AudibleAlert alert;
-  const char* uri;
-  bool loop;
-} sound_file;
-
-extern "C"{
-#include "slplay.h"
-}
-
-int last_volume = 0;
-
-void set_volume(int volume) {
-  if (last_volume != volume) {
-    char volume_change_cmd[64];
-    sprintf(volume_change_cmd, "service call audio 3 i32 3 i32 %d i32 1 &", volume);
+struct Sound::Player {
+  SLObjectItf player;
+  SLPlayItf playInterface;
+};
 
-    // 5 second timeout at 60fps
-    int volume_changed = system(volume_change_cmd);
-    last_volume = volume;
+bool Sound::init(int volumn) {
+  SLEngineOption engineOptions[] = {{SL_ENGINEOPTION_THREADSAFE, SL_BOOLEAN_TRUE}};
+  const SLInterfaceID ids[1] = {SL_IID_VOLUME};
+  const SLboolean req[1] = {SL_BOOLEAN_FALSE};
+  SLEngineItf engineInterface = NULL;
+  ReturnOnError(slCreateEngine(&engine_, 1, engineOptions, 0, NULL, NULL), "Failed to create OpenSL engine");
+  ReturnOnError((*engine_)->Realize(engine_, SL_BOOLEAN_FALSE), "Failed to realize OpenSL engine");
+  ReturnOnError((*engine_)->GetInterface(engine_, SL_IID_ENGINE, &engineInterface), "Failed to realize OpenSL engine");
+  ReturnOnError((*engineInterface)->CreateOutputMix(engineInterface, &outputMix_, 1, ids, req), "Failed to create output mix");
+  ReturnOnError((*outputMix_)->Realize(outputMix_, SL_BOOLEAN_FALSE), "Failed to realize output mix");
+
+  for (auto &kv : sound_map) {
+    if (!createPlayer(engineInterface, kv.first, kv.second)) {
+      return false;
+    }
   }
+  setVolume(volumn);
+  return true;
 }
 
+bool Sound::createPlayer(SLEngineItf engineInterface, AudibleAlert alert, const char* uri) {
```

Comment:
Why not use the uri from the map?

Reply:
I have removed the function createPlayer to make interface clear.is that ok?